### DISCUSSION TITLE
Production Docker Swarm in a world with DMZs

### DISCUSSION THREAD

**Discussion by chrisbecke** [BretFisher/ama#202] [Q&A]
Do you have any in depth talks, training or resources that deal with how to architect production swarm clusters.

For some reason I only discovered recently the vast archive of resources with the name Bret Fisher on them, and the GOTO; 2018 DockerCon talk on going Docker, Kubernetes and Swarm in production is something I need to have watched 2 years ago.
Prior to this talk nothing gave anything other than the most hand-wavy guidance on manager nodes vs workers at various scales and it was the first talk ever that mentioned an issue that plagues me: Gven a DMZ and an API zone ive been entirely without any guidance on best practice ways to split one or more clusters over that boundary - until now.

But the mention is brief, and now I am thirsty for more information

And confused what to do about traefik, that I have been deploying to manager nodes for my ingress as it needs them.

And how I convince myself and Security/Complaiance that all our mssql and postgres instances are "ok" living on dmz hosted docker hosts.

**Comment by BretFisher** [comment]
Note, that in 2021 if you need to deal with complex business security requirements and any sort of compliance at all, you'll likely just want to move to Kubernetes. Nearly all my Swarm resources are 2016-2018, and the ecosystem of tools and community support for Swarm is a tiny fraction of K8s at this point, so you'll keep running into the issues you've mentioned above.  Most 3rd party tools that previously supported Swarm have either died or moved on to only supporting K8s. What Swarm makes up for in ease of setup and management, it will limit you when you try to use persistent multi-node storage, advanced networking setups, and integrations with 3rd party tools (secrets, monitoring, auto-deploy and auto-scale) etc. Swarm is still great for personal projects or smaller projects with simpler requirements and no big need for redundant persistent data.

But if you want to give it a shot, there are still people that find it useful in specific setups:

1. A Swarm cluster with a DMZ: You would usually have, for a single set of applications in production, a single swarm cluster spanning networks. The DMZ nodes would not be managers. If the DMZ nodes were compromised, the attacker couldn't just hop to backend nodes or control the swarm. I generally always recommend having managers on a network segment that doesn't have ingress client connections to it, and use my [firewall port list](https://www.bretfisher.com/docker-swarm-firewall-ports/) to ensure the Swarm ports are open between all nodes.
2. Other resources on Swarm I recommend for solving production cluster design and use:

- https://dockerswarm.rocks
- https://github.com/swarmstack/swarmstack
- https://blog.viktoradam.net/tag/swarm/
- https://blog.sunekeller.dk/tags-archive/#swarm
- https://github.com/swarm-pack/swarm-sync
- https://github.com/stefanprodan/swarmprom
- https://github.com/djmaze/shepherd
- https://gitlab.com/n0r1sk/bosnd
- https://github.com/flavioaiello/swarm-router
- https://github.com/crazy-max/swarm-cronjob

**Comment by chrisbecke** [comment]
I find myself on the one hand, in the regretful situation of being a developer, now devops, supporting a docker cluster while my org makes its very ... deliberate way to a container strategy, that is probably K8s. My own team started with docker, then docker swarm, some time ago with the understanding that it was an interim thing we would support ourselves and, by the time we were ready to go live, the orgs container solution would be ready. it wasn't. And we've been productionizing our swarm ever since.
On the other hand, swarm is actually awesome and I really enjoy it. Docker and DevOps generally is very different from the C++ work I started in.
At any rate, thanks for amazingly helpful stuff. Looking forwards to following all your K82s stuff as soon as my Company actually commits.

**Comment by chrisbecke** [comment]
For persistent storage, my IT department was happy with deploying glusterfs to their own SANs or whatever. docker volume drivers that work nicely are hens teeth right now, but I wanted one that would create volumes for stacks properly so until the docker swarm oci support goes live, I wrote this https://github.com/chrisbecke/glusterfs-volume to mount gfs volumes and host docker volumes on them.

we use docker secrets, monitoring is done via prometheus, loki and grafana, auto-deploy? we do CD to swarm(s) using `docker -H tcp:// stack deploy` to a mtls protected secure port from gitlab pipelines. And auto-scale in our on-prem setup is a non flyer anyway and currently involves tickets and a week long wait for a build. I'm asking for api tokens and an endpoint to the ESX I can point terraform at so I can get something a bit more immediate. It's a work in progress.
